File: suitecrm-migration/frontend/shell/src/app/core/services/case.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface Case {
  id: string;
  name: string;
  caseNumber: number;
  status: string;
  priority: string;
  type: string;
  description: string;
  resolution: string;
  accountId: string;
  accountName: string;
  contactId: string;
  contactName: string;
  assignedUserId: string;
  dateEntered: string;
  dateModified: string;
}

export interface CaseUpdate {
  id: string;
  caseId: string;
  name: string;
  description: string;
  internal: boolean;
  createdBy: string;
  dateEntered: string;
}

export interface Bug {
  id: string;
  bugNumber: number;
  name: string;
  status: string;
  priority: string;
  type: string;
  description: string;
  resolution: string;
  foundInRelease: string;
  fixedInRelease: string;
}

@Injectable({ providedIn: 'root' })
export class CaseService {
  private caseUrl = `${environment.apiGatewayUrl}/api/v1/cases`;
  private bugUrl = `${environment.apiGatewayUrl}/api/v1/bugs`;

  constructor(private http: HttpClient) {}

  getCases(page = 0, size = 20): Observable<any> {
    const params = new HttpParams().set('page', page.toString()).set('size', size.toString());
    return this.http.get(this.caseUrl, { params });
  }

  getCase(id: string): Observable<Case> {
    return this.http.get<Case>(`${this.caseUrl}/${id}`);
  }

  createCase(caseData: Partial<Case>): Observable<Case> {
    return this.http.post<Case>(this.caseUrl, caseData);
  }

  updateCase(id: string, caseData: Partial<Case>): Observable<Case> {
    return this.http.put<Case>(`${this.caseUrl}/${id}`, caseData);
  }

  deleteCase(id: string): Observable<void> {
    return this.http.delete<void>(`${this.caseUrl}/${id}`);
  }

  getCaseUpdates(caseId: string): Observable<CaseUpdate[]> {
    return this.http.get<CaseUpdate[]>(`${this.caseUrl}/${caseId}/updates`);
  }

  addCaseUpdate(caseId: string, update: Partial<CaseUpdate>): Observable<CaseUpdate> {
    return this.http.post<CaseUpdate>(`${this.caseUrl}/${caseId}/updates`, update);
  }

  getBugs(page = 0, size = 20): Observable<any> {
    const params = new HttpParams().set('page', page.toString()).set('size', size.toString());
    return this.http.get(this.bugUrl, { params });
  }

  getBug(id: string): Observable<Bug> {
    return this.http.get<Bug>(`${this.bugUrl}/${id}`);
  }
}

File: suitecrm-migration/backend/case-service/src/main/java/com/suitecrm/cases/service/CaseService.java
package com.suitecrm.cases.service;

import com.suitecrm.cases.dto.*;
import com.suitecrm.cases.entity.CaseEvent;
import com.suitecrm.cases.entity.CaseUpdate;
import com.suitecrm.cases.entity.SupportCase;
import com.suitecrm.cases.repository.CaseEventRepository;
import com.suitecrm.cases.repository.CaseUpdateRepository;
import com.suitecrm.cases.repository.SupportCaseRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional
public class CaseService {

    private final SupportCaseRepository supportCaseRepository;
    private final CaseEventRepository caseEventRepository;
    private final CaseUpdateRepository caseUpdateRepository;

    @Transactional(readOnly = true)
    public Page<CaseDto> listCases(int page, int size, String sortBy, String sortDir, String search) {
        Sort sort = sortDir.equalsIgnoreCase("asc") ? Sort.by(sortBy).ascending() : Sort.by(sortBy).descending();
        Pageable pageable = PageRequest.of(page, size, sort);
        Page<SupportCase> cases;
        if (search != null && !search.isBlank()) {
            cases = supportCaseRepository.searchCases(search, pageable);
        } else {
            cases = supportCaseRepository.findByDeletedFalse(pageable);
        }
        return cases.map(this::toDto);
    }

    @Transactional(readOnly = true)
    public CaseDto getCase(UUID id) {
        SupportCase sc = supportCaseRepository.findByIdAndDeletedFalse(id)
                .orElseThrow(() -> new RuntimeException("Case not found with id: " + id));
        return toDto(sc);
    }

    public CaseDto createCase(CaseCreateRequest request, UUID createdBy) {
        log.info("Creating case: name={}", request.getName());
        SupportCase sc = SupportCase.builder()
                .name(request.getName())
                .status(request.getStatus() != null ? request.getStatus() : "New")
                .priority(request.getPriority() != null ? request.getPriority() : "Medium")
                .type(request.getType())
                .description(request.getDescription())
                .resolution(request.getResolution())
                .accountId(request.getAccountId())
                .assignedUserId(request.getAssignedUserId())
                .createdBy(createdBy)
                .build();
        SupportCase saved = supportCaseRepository.save(sc);

        CaseEvent event = CaseEvent.builder()
                .caseId(saved.getId())
                .eventType("CREATED")
                .description("Case created")
                .createdBy(createdBy)
                .build();
        caseEventRepository.save(event);

        log.info("Created case: id={}, number={}", saved.getId(), saved.getCaseNumber());
        return toDto(saved);
    }

    public CaseDto updateCase(UUID id, CaseCreateRequest request, UUID updatedBy) {
        SupportCase existing = supportCaseRepository.findByIdAndDeletedFalse(id)
                .orElseThrow(() -> new RuntimeException("Case not found with id: " + id));
        String oldStatus = existing.getStatus();
        existing.setName(request.getName());
        existing.setStatus(request.getStatus());
        existing.setPriority(request.getPriority());
        existing.setType(request.getType());
        existing.setDescription(request.getDescription());
        existing.setResolution(request.getResolution());
        existing.setAssignedUserId(request.getAssignedUserId());
        SupportCase saved = supportCaseRepository.save(existing);

        if (!oldStatus.equals(request.getStatus())) {
            CaseEvent event = CaseEvent.builder()
                    .caseId(saved.getId())
                    .eventType("STATUS_CHANGED")
                    .description("Status changed from " + oldStatus + " to " + request.getStatus())
                    .createdBy(updatedBy)
                    .build();
            caseEventRepository.save(event);
        }

        return toDto(saved);
    }

    public void addCaseUpdate(UUID caseId, String description, boolean internal, UUID createdBy) {
        CaseUpdate update = CaseUpdate.builder()
                .caseId(caseId)
                .description(description)
                .internal(internal)
                .createdBy(createdBy)
                .build();
        caseUpdateRepository.save(update);
    }

    @Transactional(readOnly = true)
    public List<CaseUpdate> getCaseUpdates(UUID caseId) {
        return caseUpdateRepository.findByCaseIdAndDeletedFalseOrderByDateEnteredDesc(caseId);
    }

    @Transactional(readOnly = true)
    public List<CaseEvent> getCaseEvents(UUID caseId) {
        return caseEventRepository.findByCaseIdAndDeletedFalseOrderByDateEnteredDesc(caseId);
    }

    public void deleteCase(UUID id) {
        SupportCase sc = supportCaseRepository.findByIdAndDeletedFalse(id)
                .orElseThrow(() -> new RuntimeException("Case not found with id: " + id));
        sc.setDeleted(true);
        supportCaseRepository.save(sc);
    }

    private CaseDto toDto(SupportCase entity) {
        return CaseDto.builder()
                .id(entity.getId())
                .name(entity.getName())
                .caseNumber(entity.getCaseNumber())
                .status(entity.getStatus())
                .priority(entity.getPriority())
                .type(entity.getType())
                .description(entity.getDescription())
                .resolution(entity.getResolution())
                .accountId(entity.getAccountId())
                .assignedUserId(entity.getAssignedUserId())
                .createdBy(entity.getCreatedBy())
                .dateEntered(entity.getDateEntered())
                .dateModified(entity.getDateModified())
                .build();
    }
}
